src/core/manager.c

```diff
@@ -2288,16 +2288,19 @@ static int manager_dispatch_cgroups_agent_fd(sd_event_source *source, int fd, ui
         return 0;
 }
 
-static bool manager_process_barrier_fd(const char *buf, FDSet *fds) {
-        assert(buf);
+static bool manager_process_barrier_fd(char * const *tags, FDSet *fds) {
 
         /* nothing else must be sent when using BARRIER=1 */
-        if (STR_IN_SET(buf, "BARRIER=1", "BARRIER=1\n")) {
-                if (fdset_size(fds) != 1)
-                        log_warning("Got incorrect number of fds with BARRIER=1, closing them.");
+        if (strv_contains(tags, "BARRIER=1")) {
+                if (strv_length(tags) == 1) {
```

Comment:
you iterate through the array twice here. maybe:

```c
if (strv_equal(tags, STRV_MAKE("BARRIER=1")) {
        …
} else if (strv_contains(tags, "BARRIER=1")) {
        …
}
```

that way you iterate through the array only once.

Reply:
Since the reception of "" BARRIER = 1 "` is less frequent (currently nobody uses it, but statically, it will always be less frequent), we can write in the opposite direction:
```
        if (strv_contains(tags, "BARRIER=1")) {
                if (strv_equal(tags, STRV_MAKE("BARRIER=1")) {
```
But technically strv_length is very cheap, and faster than `strv_equal()`  (Just iterating on a array costs less than a string compare).
So even if there is a slight improvement, I do not think that it is easier to read.

Reply:
If somebody really care about performance, `(strv_length(tags) == 1)` can be replaced (in this specific case) by `(tags[1] == NULL)`
But I don't think this is necessary

Reply:
strv_equal() is O(1) time in this case where the second argument is literal, it's hard to beat that.